tools/generate_release_notes.py

```diff
@@ -1,246 +1,331 @@
+            authors.add(commit.author)
+        if commit.committer:
+            reviewers.add(commit.committer)
 
-args = parser.parse_args()
+    for pull in pull_requests:
+        for review in pull.get_reviews():
+            if review.user:
+                reviewers.add(review.user)
 
-for tag in repository.get_tags():
-    if tag.name == args.from_commit:
-        previous_tag = tag
-        break
-else:
-    raise RuntimeError(f"Desired tag ({args.from_commit}) not found")
+    return authors, reviewers
 
-# For some reason, go get the github commit from the commit to get
-# the correct date
-github_commit = previous_tag.commit.commit
-previous_tag_date = datetime.strptime(
-    github_commit.last_modified, "%a, %d %b %Y %H:%M:%S %Z"
-)
 
+@dataclass(frozen=True, kw_only=True)
+class MdFormatter:
+    """Format release notes in Markdown from PRs, authors and reviewers."""
 
-all_commits = list(
-    tqdm(
-        repository.get_commits(sha=args.to_commit, since=previous_tag_date),
-        desc=f"Getting all commits between {args.from_commit} " f"and {args.to_commit}",
-    )
-)
-all_hashes = {c.sha for c in all_commits}
+    pull_requests: set[PullRequest]
+    authors: set[NamedUser]
+    reviewers: set[NamedUser]
 
-authors = set()
-reviewers = set()
-committers = set()
-users = dict()  # keep track of known usernames
+    version: str = "x.y.z"
+    title_template: str = "scikit-image {version} release notes"
+    intro_template: str = """
+We're happy to announce the release of scikit-image {version}!
+scikit-image is an image processing toolbox for SciPy that includes algorithms
+for segmentation, geometric transformations, color space manipulation,
+analysis, filtering, morphology, feature detection, and more.
 
+For more information, examples, and documentation, please visit our website:
+https://scikit-image.org
 
-def find_author_info(commit):
-    """Return committer and author of a commit.
+"""
+    label_section_map: tuple[str, str] = (
+        (":trophy: type: Highlight", "Highlights"),
+        (":baby: type: New feature", "New Features"),
+        (":fast_forward: type: Enhancement", "Enhancements"),
+        (":chart_with_upwards_trend: type: Performance", "Performance"),
+        (":adhesive_bandage: type: Bug fix", "Bug Fixes"),
+        (":scroll: type: API", "API Changes"),
+        (":wrench: type: Maintenance", "Maintenance"),
+        (":page_facing_up: type: Documentation", "Documentation"),
+        (":robot: type: Infrastructure", "Infrastructure"),
+    )
+    ignored_user_logins: tuple[str] = ("web-flow",)
+    pr_summary_regex = re.compile(
+        r"^```release-note\s*(?P<summary>[\s\S]*?\w[\s\S]*?)\s*^```", flags=re.MULTILINE
+    )
 
-    Parameters
-    ----------
-    commit : Github commit
-        The commit to query.
+    @property
+    def intro(self):
+        return self.intro_template.format(version=self.version)
+
+    def __str__(self) -> str:
+        return "".join(self.iter_lines())
```

Comment:
Why not implement `__iter__`?

Reply:
Thought about it, but I always feel like they are sometimes a worse option than an expressive method or attribute. `__str__`, `__iter__` and such rely on the object's name being expressive, e.g. `list(formatter.iter_lines())` seems a lot more expressive to me than `list(formatter)`.

It's a gut feeling though. I opted to make everyone happy by providing all options in https://github.com/scikit-image/scikit-image/pull/6961/commits/3b788d350edc4204f33dee27fdbafd8f54ff5255. :smile: